types/index.d.ts

```diff
@@ -319,13 +319,12 @@ declare namespace math {
   type OperatorNodeFn = keyof OperatorNodeMap
 
   interface OperatorNode<
-    TOp extends OperatorNodeMap[TFn] = never,
-    TFn extends OperatorNodeFn = never,
+    TFn extends OperatorNodeFn = OperatorNodeFn,
```

Comment:
I should have isolated this issue more, but I am pretty sure that...

In the old setup, something like `new OperatorNode("+", "addition", ...)" was never going to be assignable to  be assignable to a MathNode, because:
-  MathNode union looked like `NodeType1 | NodeType2 ... | OperatorNode | ...`, i.e., OperatorNode was using its defaults in MathNode union...
- and OperatorNode default included `op: never`, so nothing was ever going to be assignable to OperatorNode.

To remedy this, I originally wanted:
```ts
  interface OperatorNode<
    TOp extends OperatorNodeMap[TFn] = OperatorNodeMap[TFn],
    TFn extends OperatorNodeFn = OperatorNodeFn,
    ...>{...}
```
But TS complained later parameters TFn can't be used as defaults for earlier parameters. So originally I was going to swap the order of the two, but then I just got rid of TOp.


**NOTE:** Fixing this issue may not have been entirely necessary to get the discriminated union stuff to work, I'm not 100% sure, and I need to run soon so can't try to isolate it more right now.

Reply:
>In the old setup, something like `new OperatorNode("+", "addition", ...)" was never going to be assignable to be assignable to a MathNode, because:

Do you have a reproducible example of this? Maybe I'm misunderstanding but assigning `new OperatorNode("+", "addition", ...)` to `MathNode`  seems to work for me (at least in test/typescript-tests/testTypes.ts):

<img width="586" alt="image" src="https://user-images.githubusercontent.com/64985/199555961-6f9d3fc5-98c8-41e3-9bd6-fc89d54c3eef.png">

> But TS complained later parameters TFn can't be used as defaults for earlier parameters. So originally I was going to swap the order of the two, but then I just got rid of TOp.

I think it's going to be a fairly big breaking change to change the generics for `OperatorNode` so I'd suggest we try and instead find a way around whatever issue there is here.

Reply:
tldr: I agree we should not change the OperatorNode signature in this PR. Instead we should use `OperatorNode<OperatorNodeOp, OperatorNodeFn>` in the MathNodes interface. I do think there's an issue with the type definition for OperatorNode, but best not to change it here.

---

The test you screenshotted above will fail on your branch; I agree it passes on master. Simpler version:
```ts
// passes on MathJs 11.2.1, (last version that used discriminated union)
// passes on MathJs ^11.3.0 (when discriminated union was removed)
// fails on goodproblems/types/revert-to-discr-union-for-nodes 
const node1: MathNode = new OperatorNode("+, "add", [new ConstantNode(1), new ConstantNode(1)])
```

Looking back at the old typedefs, the old discriminated union was:
```ts
type MathNode =  ... | ... |  OperatorNode<OperatorNodeOp, OperatorNodeFn> | ... ;
```

So I think we should go back to using `OperatorNode<OperatorNodeOp, OperatorNodeFn>` not `OperatorNode` (with its default arguments). That's definitely a more focused change than changing the type parameters.
---

But.. I do think that there is an issue with OperatorNode's type definition (a separate issue that needs not be handled here). Namely, its default arguments seem useless to me. No OperatorNode instance can ever be assignable `OperatorNode` type (with its default args): Currently on master:
<img width="400" alt="Screen Shot 2022-11-02 at 9 26 36 PM" src="https://user-images.githubusercontent.com/9010790/199632009-2db3a888-6c9e-4043-bbb8-cb78fe98ec0b.png">

Better to change the type signature or remove the defaults, imo. _But, again, I agree: best to avoid that here._

Reply:
>The test you screenshotted above will fail on your branch; I agree it passes on master. Simpler version:

I was on that wrong branch, you're correct. 🙄 

>So I think we should go back to using OperatorNode<OperatorNodeOp, OperatorNodeFn> not OperatorNode (with its default arguments). That's definitely a more focused change than changing the type parameters.

Awesome, looks good 👍🏻 

> But.. I do think that there is an issue with OperatorNode's type definition (a separate issue that needs not be handled here). Namely, its default arguments seem useless to me. No OperatorNode instance can ever be assignable OperatorNode type (with its default args): Currently on master:

This was the reason for using `never`: https://github.com/josdejong/mathjs/pull/2576#issuecomment-1139760122, not sure it's still a good reason. Happy to check out a PR which changes this.